File: src/algo/leetcode/search/binarySearch.cpp

bool binarySearch(int arr[], int length, int item) {
    int l = 0, r = length - 1,  mid = 0;
    while( l <= r) {
        mid = (l + r) / 2/
        if(item == arr[mid]) {
            return true;
        } else if( item < arr[mid]) {
            r = mid - 1;
        } else {
            l + mid + 1;
        }
    }
    return false;
}

int firstOccurrence(int arr[], int left, int right, int item) {
    int mid = 0;
    while(right - left > 1) {
        mid = (left + right) / 2;
        if(arr[mid] >= item) {
            right = mid;
        } else {
            left = mid;
        }
    }
    return right;
}

File: src/go/binarySearch.go
package main

import (
	"fmt"
)

func binarySearch(item int, arr []int) int {
	if arr == nil || len(arr) <= 0 {
		return -1
	}

	lo, hi := 0, len(arr)
	while(low <= hi) {
		mid = low + (hi - lo) / 2
		if mid == item {
			return mid
		} 
	}

	return 0;

}

func main() {
	arr := []int{1,2,3,9,8,7,10,22,9,10}
	item := 1
	
	index := binarySearch(item, arr)
	fmt.Println(index)
}